{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"name":"python","version":"3.10.14","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"},"kaggle":{"accelerator":"none","dataSources":[{"sourceId":9266991,"sourceType":"datasetVersion","datasetId":5484830}],"dockerImageVersionId":30761,"isInternetEnabled":true,"language":"python","sourceType":"notebook","isGpuEnabled":false}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"markdown","source":"### Overfing","metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19"}},{"cell_type":"code","source":"import pandas as pd\n\ndf = pd.read_parquet('/kaggle/input/cleaned-mutations-phd/df.parquet.hugo.snp.ins.del.all.qzip')","metadata":{"execution":{"iopub.status.busy":"2024-08-28T14:06:07.061999Z","iopub.execute_input":"2024-08-28T14:06:07.062618Z","iopub.status.idle":"2024-08-28T14:06:11.874750Z","shell.execute_reply.started":"2024-08-28T14:06:07.062558Z","shell.execute_reply":"2024-08-28T14:06:11.873596Z"},"trusted":true},"execution_count":3,"outputs":[]},{"cell_type":"code","source":"df = df.dropna(subset=['Gene', 'Exon'])","metadata":{"execution":{"iopub.status.busy":"2024-08-28T14:08:33.958496Z","iopub.execute_input":"2024-08-28T14:08:33.959730Z","iopub.status.idle":"2024-08-28T14:08:35.791931Z","shell.execute_reply.started":"2024-08-28T14:08:33.959670Z","shell.execute_reply":"2024-08-28T14:08:35.790443Z"},"trusted":true},"execution_count":7,"outputs":[]},{"cell_type":"code","source":"df","metadata":{"execution":{"iopub.status.busy":"2024-08-28T14:08:49.186853Z","iopub.execute_input":"2024-08-28T14:08:49.187435Z","iopub.status.idle":"2024-08-28T14:08:49.218181Z","shell.execute_reply.started":"2024-08-28T14:08:49.187381Z","shell.execute_reply":"2024-08-28T14:08:49.216223Z"},"trusted":true},"execution_count":8,"outputs":[{"execution_count":8,"output_type":"execute_result","data":{"text/plain":"                                          primary_site  \\\n0        Hematopoietic and reticuloendothelial systems   \n1        Hematopoietic and reticuloendothelial systems   \n2        Hematopoietic and reticuloendothelial systems   \n4        Hematopoietic and reticuloendothelial systems   \n5        Hematopoietic and reticuloendothelial systems   \n...                                                ...   \n3570295                                   Corpus uteri   \n3570296                              Bronchus and lung   \n3570297                              Bronchus and lung   \n3570298                                   Corpus uteri   \n3570299                                        Stomach   \n\n                                  file_gdc_id project_short_name Variant_Type  \\\n0        5c0a24b5-ed57-4ac8-b7c7-6b839622ca47  BEATAML1.0-COHORT          DEL   \n1        ad0dda7c-8df4-4160-a1a7-6657a7c4b74b  BEATAML1.0-COHORT          DEL   \n2        ff14dfb7-a8ae-444a-8389-f189dfbb1f86  BEATAML1.0-COHORT          DEL   \n4        9be6cc7f-3f0c-49c8-b860-b333aa76dff3  BEATAML1.0-COHORT          DEL   \n5        95a5341f-4ad8-4d58-83d5-3cd5515ce210  BEATAML1.0-COHORT          DEL   \n...                                       ...                ...          ...   \n3570295  c676f7e1-de13-4c16-b106-a611228881b3          TCGA-UCEC          SNP   \n3570296  76dce8cc-38ff-4ddf-8272-0e405d82f63a          TCGA-LUAD          SNP   \n3570297  ebe634ff-b960-4474-8c6c-1451ad866f6b          TCGA-LUAD          SNP   \n3570298  a34c7bfa-9a11-4974-812f-dc682482d39b          TCGA-UCEC          SNP   \n3570299  73b4138f-ac42-41a6-b560-7ee53388c02d          TCGA-STAD          SNP   \n\n        Variant_Classification        Reference_Allele  \\\n0              Frame_Shift_Del  GCGGCAGGTGGGGCGGGAGGCT   \n1                 In_Frame_Del                     AGT   \n2              Frame_Shift_Del                 CGTTCCA   \n4              Frame_Shift_Del                      AC   \n5                  Splice_Site            TGGGACATCTGC   \n...                        ...                     ...   \n3570295      Missense_Mutation                       T   \n3570296      Missense_Mutation                       T   \n3570297      Missense_Mutation                       T   \n3570298                 Silent                       T   \n3570299      Missense_Mutation                       T   \n\n              Tumor_Seq_Allele1 Tumor_Seq_Allele2             Gene  \\\n0        GCGGCAGGTGGGGCGGGAGGCT                 -  ENSG00000245848   \n1                           AGT                 -  ENSG00000134480   \n2                       CGTTCCA                 -  ENSG00000108312   \n4                            AC                 -  ENSG00000164754   \n5                  TGGGACATCTGC         GCCCCAGGT  ENSG00000181555   \n...                         ...               ...              ...   \n3570295                       T                 C  ENSG00000173085   \n3570296                       T                 C  ENSG00000157219   \n3570297                       T                 C  ENSG00000111077   \n3570298                       T                 C  ENSG00000197121   \n3570299                       T                 G  ENSG00000171951   \n\n                 Feature Hugo_Symbol   Exon      Start        End  \\\n0        ENST00000498907       CEBPA    1/1   33302272   33302293   \n1        ENST00000256897        CCNH    7/9   87399494   87399496   \n2        ENST00000302904        UBTF  13/21   44210812   44210818   \n4        ENST00000297338       RAD21  10/14  116852694  116852695   \n5        ENST00000409792       SETD2  20/21   47017732   47017743   \n...                  ...         ...    ...        ...        ...   \n3570295  ENST00000311469        COQ2    6/7   83267629   83267629   \n3570296  ENST00000287907       HTR5A    1/2  155071267  155071267   \n3570297  ENST00000314250        TNS2  12/29   53057611   53057611   \n3570298  ENST00000354764       PGAP1   4/27  196912985  196912985   \n3570299  ENST00000305409        SCG2    2/2  223599137  223599137   \n\n            Genomic_Bin Strand  \\\n0        chr19_33000000      +   \n1         chr5_87000000      +   \n2        chr17_44000000      +   \n4        chr8_116000000      +   \n5         chr3_47000000      +   \n...                 ...    ...   \n3570295   chr4_83000000      +   \n3570296  chr7_155000000      +   \n3570297  chr12_53000000      +   \n3570298  chr2_196000000      +   \n3570299  chr2_223000000      +   \n\n                                                   dataset Chromosome  \\\n0        `isb-cgc-bq.BEATAML1_0.masked_somatic_mutation...      chr19   \n1        `isb-cgc-bq.BEATAML1_0.masked_somatic_mutation...       chr5   \n2        `isb-cgc-bq.BEATAML1_0.masked_somatic_mutation...      chr17   \n4        `isb-cgc-bq.BEATAML1_0.masked_somatic_mutation...       chr8   \n5        `isb-cgc-bq.BEATAML1_0.masked_somatic_mutation...       chr3   \n...                                                    ...        ...   \n3570295  `isb-cgc-bq.TCGA.masked_somatic_mutation_hg38_...       chr4   \n3570296  `isb-cgc-bq.TCGA.masked_somatic_mutation_hg38_...       chr7   \n3570297  `isb-cgc-bq.TCGA.masked_somatic_mutation_hg38_...      chr12   \n3570298  `isb-cgc-bq.TCGA.masked_somatic_mutation_hg38_...       chr2   \n3570299  `isb-cgc-bq.TCGA.masked_somatic_mutation_hg38_...       chr2   \n\n        Mutational_Motif  \n0                G[del]T  \n1                A[del]T  \n2                T[del]A  \n4                G[del]C  \n5                C[del]C  \n...                  ...  \n3570295          G[T>C]T  \n3570296          C[T>C]T  \n3570297          A[T>C]T  \n3570298          A[T>C]T  \n3570299          C[T>G]T  \n\n[3408502 rows x 19 columns]","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>primary_site</th>\n      <th>file_gdc_id</th>\n      <th>project_short_name</th>\n      <th>Variant_Type</th>\n      <th>Variant_Classification</th>\n      <th>Reference_Allele</th>\n      <th>Tumor_Seq_Allele1</th>\n      <th>Tumor_Seq_Allele2</th>\n      <th>Gene</th>\n      <th>Feature</th>\n      <th>Hugo_Symbol</th>\n      <th>Exon</th>\n      <th>Start</th>\n      <th>End</th>\n      <th>Genomic_Bin</th>\n      <th>Strand</th>\n      <th>dataset</th>\n      <th>Chromosome</th>\n      <th>Mutational_Motif</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>Hematopoietic and reticuloendothelial systems</td>\n      <td>5c0a24b5-ed57-4ac8-b7c7-6b839622ca47</td>\n      <td>BEATAML1.0-COHORT</td>\n      <td>DEL</td>\n      <td>Frame_Shift_Del</td>\n      <td>GCGGCAGGTGGGGCGGGAGGCT</td>\n      <td>GCGGCAGGTGGGGCGGGAGGCT</td>\n      <td>-</td>\n      <td>ENSG00000245848</td>\n      <td>ENST00000498907</td>\n      <td>CEBPA</td>\n      <td>1/1</td>\n      <td>33302272</td>\n      <td>33302293</td>\n      <td>chr19_33000000</td>\n      <td>+</td>\n      <td>`isb-cgc-bq.BEATAML1_0.masked_somatic_mutation...</td>\n      <td>chr19</td>\n      <td>G[del]T</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>Hematopoietic and reticuloendothelial systems</td>\n      <td>ad0dda7c-8df4-4160-a1a7-6657a7c4b74b</td>\n      <td>BEATAML1.0-COHORT</td>\n      <td>DEL</td>\n      <td>In_Frame_Del</td>\n      <td>AGT</td>\n      <td>AGT</td>\n      <td>-</td>\n      <td>ENSG00000134480</td>\n      <td>ENST00000256897</td>\n      <td>CCNH</td>\n      <td>7/9</td>\n      <td>87399494</td>\n      <td>87399496</td>\n      <td>chr5_87000000</td>\n      <td>+</td>\n      <td>`isb-cgc-bq.BEATAML1_0.masked_somatic_mutation...</td>\n      <td>chr5</td>\n      <td>A[del]T</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>Hematopoietic and reticuloendothelial systems</td>\n      <td>ff14dfb7-a8ae-444a-8389-f189dfbb1f86</td>\n      <td>BEATAML1.0-COHORT</td>\n      <td>DEL</td>\n      <td>Frame_Shift_Del</td>\n      <td>CGTTCCA</td>\n      <td>CGTTCCA</td>\n      <td>-</td>\n      <td>ENSG00000108312</td>\n      <td>ENST00000302904</td>\n      <td>UBTF</td>\n      <td>13/21</td>\n      <td>44210812</td>\n      <td>44210818</td>\n      <td>chr17_44000000</td>\n      <td>+</td>\n      <td>`isb-cgc-bq.BEATAML1_0.masked_somatic_mutation...</td>\n      <td>chr17</td>\n      <td>T[del]A</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>Hematopoietic and reticuloendothelial systems</td>\n      <td>9be6cc7f-3f0c-49c8-b860-b333aa76dff3</td>\n      <td>BEATAML1.0-COHORT</td>\n      <td>DEL</td>\n      <td>Frame_Shift_Del</td>\n      <td>AC</td>\n      <td>AC</td>\n      <td>-</td>\n      <td>ENSG00000164754</td>\n      <td>ENST00000297338</td>\n      <td>RAD21</td>\n      <td>10/14</td>\n      <td>116852694</td>\n      <td>116852695</td>\n      <td>chr8_116000000</td>\n      <td>+</td>\n      <td>`isb-cgc-bq.BEATAML1_0.masked_somatic_mutation...</td>\n      <td>chr8</td>\n      <td>G[del]C</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>Hematopoietic and reticuloendothelial systems</td>\n      <td>95a5341f-4ad8-4d58-83d5-3cd5515ce210</td>\n      <td>BEATAML1.0-COHORT</td>\n      <td>DEL</td>\n      <td>Splice_Site</td>\n      <td>TGGGACATCTGC</td>\n      <td>TGGGACATCTGC</td>\n      <td>GCCCCAGGT</td>\n      <td>ENSG00000181555</td>\n      <td>ENST00000409792</td>\n      <td>SETD2</td>\n      <td>20/21</td>\n      <td>47017732</td>\n      <td>47017743</td>\n      <td>chr3_47000000</td>\n      <td>+</td>\n      <td>`isb-cgc-bq.BEATAML1_0.masked_somatic_mutation...</td>\n      <td>chr3</td>\n      <td>C[del]C</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>3570295</th>\n      <td>Corpus uteri</td>\n      <td>c676f7e1-de13-4c16-b106-a611228881b3</td>\n      <td>TCGA-UCEC</td>\n      <td>SNP</td>\n      <td>Missense_Mutation</td>\n      <td>T</td>\n      <td>T</td>\n      <td>C</td>\n      <td>ENSG00000173085</td>\n      <td>ENST00000311469</td>\n      <td>COQ2</td>\n      <td>6/7</td>\n      <td>83267629</td>\n      <td>83267629</td>\n      <td>chr4_83000000</td>\n      <td>+</td>\n      <td>`isb-cgc-bq.TCGA.masked_somatic_mutation_hg38_...</td>\n      <td>chr4</td>\n      <td>G[T&gt;C]T</td>\n    </tr>\n    <tr>\n      <th>3570296</th>\n      <td>Bronchus and lung</td>\n      <td>76dce8cc-38ff-4ddf-8272-0e405d82f63a</td>\n      <td>TCGA-LUAD</td>\n      <td>SNP</td>\n      <td>Missense_Mutation</td>\n      <td>T</td>\n      <td>T</td>\n      <td>C</td>\n      <td>ENSG00000157219</td>\n      <td>ENST00000287907</td>\n      <td>HTR5A</td>\n      <td>1/2</td>\n      <td>155071267</td>\n      <td>155071267</td>\n      <td>chr7_155000000</td>\n      <td>+</td>\n      <td>`isb-cgc-bq.TCGA.masked_somatic_mutation_hg38_...</td>\n      <td>chr7</td>\n      <td>C[T&gt;C]T</td>\n    </tr>\n    <tr>\n      <th>3570297</th>\n      <td>Bronchus and lung</td>\n      <td>ebe634ff-b960-4474-8c6c-1451ad866f6b</td>\n      <td>TCGA-LUAD</td>\n      <td>SNP</td>\n      <td>Missense_Mutation</td>\n      <td>T</td>\n      <td>T</td>\n      <td>C</td>\n      <td>ENSG00000111077</td>\n      <td>ENST00000314250</td>\n      <td>TNS2</td>\n      <td>12/29</td>\n      <td>53057611</td>\n      <td>53057611</td>\n      <td>chr12_53000000</td>\n      <td>+</td>\n      <td>`isb-cgc-bq.TCGA.masked_somatic_mutation_hg38_...</td>\n      <td>chr12</td>\n      <td>A[T&gt;C]T</td>\n    </tr>\n    <tr>\n      <th>3570298</th>\n      <td>Corpus uteri</td>\n      <td>a34c7bfa-9a11-4974-812f-dc682482d39b</td>\n      <td>TCGA-UCEC</td>\n      <td>SNP</td>\n      <td>Silent</td>\n      <td>T</td>\n      <td>T</td>\n      <td>C</td>\n      <td>ENSG00000197121</td>\n      <td>ENST00000354764</td>\n      <td>PGAP1</td>\n      <td>4/27</td>\n      <td>196912985</td>\n      <td>196912985</td>\n      <td>chr2_196000000</td>\n      <td>+</td>\n      <td>`isb-cgc-bq.TCGA.masked_somatic_mutation_hg38_...</td>\n      <td>chr2</td>\n      <td>A[T&gt;C]T</td>\n    </tr>\n    <tr>\n      <th>3570299</th>\n      <td>Stomach</td>\n      <td>73b4138f-ac42-41a6-b560-7ee53388c02d</td>\n      <td>TCGA-STAD</td>\n      <td>SNP</td>\n      <td>Missense_Mutation</td>\n      <td>T</td>\n      <td>T</td>\n      <td>G</td>\n      <td>ENSG00000171951</td>\n      <td>ENST00000305409</td>\n      <td>SCG2</td>\n      <td>2/2</td>\n      <td>223599137</td>\n      <td>223599137</td>\n      <td>chr2_223000000</td>\n      <td>+</td>\n      <td>`isb-cgc-bq.TCGA.masked_somatic_mutation_hg38_...</td>\n      <td>chr2</td>\n      <td>C[T&gt;G]T</td>\n    </tr>\n  </tbody>\n</table>\n<p>3408502 rows × 19 columns</p>\n</div>"},"metadata":{}}]},{"cell_type":"code","source":"top_sites = df['primary_site'].value_counts().nlargest(24).index\n\nfiltered_df = df[df['primary_site'].isin(top_sites)]\n\nfiltered_df['primary_site'] = filtered_df['primary_site'].astype('category').cat.codes\n\n\ngrouped_df = filtered_df.groupby('file_gdc_id').agg({\n    'Mutational_Motif': list,\n    'Genomic_Bin': list,\n    'Hugo_Symbol': list,\n    'Exon': list, \n    'Feature': list,\n    'Variant_Classification': list, \n    'primary_site': 'first'\n}).reset_index()\n","metadata":{"execution":{"iopub.status.busy":"2024-08-28T14:09:34.790489Z","iopub.execute_input":"2024-08-28T14:09:34.791106Z","iopub.status.idle":"2024-08-28T14:09:45.143007Z","shell.execute_reply.started":"2024-08-28T14:09:34.791045Z","shell.execute_reply":"2024-08-28T14:09:45.141697Z"},"trusted":true},"execution_count":9,"outputs":[{"name":"stderr","text":"/tmp/ipykernel_36/4185034200.py:5: SettingWithCopyWarning: \nA value is trying to be set on a copy of a slice from a DataFrame.\nTry using .loc[row_indexer,col_indexer] = value instead\n\nSee the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n  filtered_df['primary_site'] = filtered_df['primary_site'].astype('category').cat.codes\n","output_type":"stream"}]},{"cell_type":"code","source":"list_a = ['Mutational_Motif',\n    'Genomic_Bin',\n    'Hugo_Symbol',\n    'Exon', \n    'Feature',\n    'Variant_Classification']","metadata":{"execution":{"iopub.status.busy":"2024-08-28T14:10:26.785092Z","iopub.execute_input":"2024-08-28T14:10:26.786043Z","iopub.status.idle":"2024-08-28T14:10:26.791882Z","shell.execute_reply.started":"2024-08-28T14:10:26.785989Z","shell.execute_reply":"2024-08-28T14:10:26.790646Z"},"trusted":true},"execution_count":10,"outputs":[]},{"cell_type":"code","source":"from tensorflow.keras.preprocessing.text import Tokenizer\n\n\ntokenizers = {}\nfor column in list_a:\n    tokenizer = Tokenizer()\n    tokenizer.fit_on_texts(grouped_df[column])\n    tokenizers[column] = tokenizer\n","metadata":{"execution":{"iopub.status.busy":"2024-08-28T14:14:39.685744Z","iopub.execute_input":"2024-08-28T14:14:39.686481Z","iopub.status.idle":"2024-08-28T14:14:57.255083Z","shell.execute_reply.started":"2024-08-28T14:14:39.686411Z","shell.execute_reply":"2024-08-28T14:14:57.253418Z"},"trusted":true},"execution_count":15,"outputs":[]},{"cell_type":"code","source":"max_len = int(grouped_df['Mutational_Motif'].apply(len).quantile(0.75))","metadata":{"execution":{"iopub.status.busy":"2024-08-28T14:14:57.257280Z","iopub.execute_input":"2024-08-28T14:14:57.257730Z","iopub.status.idle":"2024-08-28T14:14:57.277374Z","shell.execute_reply.started":"2024-08-28T14:14:57.257686Z","shell.execute_reply":"2024-08-28T14:14:57.276009Z"},"trusted":true},"execution_count":16,"outputs":[]},{"cell_type":"code","source":"from tensorflow.keras.preprocessing.sequence import pad_sequences\n\nX = {}\nfor key in tokenizers.keys():\n    X[key] = pad_sequences(tokenizers[key].texts_to_sequences(grouped_df[key]), maxlen=max_len, padding='post')","metadata":{"execution":{"iopub.status.busy":"2024-08-28T14:14:57.279061Z","iopub.execute_input":"2024-08-28T14:14:57.279601Z","iopub.status.idle":"2024-08-28T14:15:09.626159Z","shell.execute_reply.started":"2024-08-28T14:14:57.279540Z","shell.execute_reply":"2024-08-28T14:15:09.624694Z"},"trusted":true},"execution_count":17,"outputs":[]},{"cell_type":"code","source":"tokenizers","metadata":{"execution":{"iopub.status.busy":"2024-08-28T14:31:29.221659Z","iopub.execute_input":"2024-08-28T14:31:29.222158Z","iopub.status.idle":"2024-08-28T14:31:29.230389Z","shell.execute_reply.started":"2024-08-28T14:31:29.222114Z","shell.execute_reply":"2024-08-28T14:31:29.228987Z"},"trusted":true},"execution_count":32,"outputs":[{"execution_count":32,"output_type":"execute_result","data":{"text/plain":"{'Mutational_Motif': <keras.src.legacy.preprocessing.text.Tokenizer at 0x7bfa572d5db0>,\n 'Genomic_Bin': <keras.src.legacy.preprocessing.text.Tokenizer at 0x7bfa56900b80>,\n 'Hugo_Symbol': <keras.src.legacy.preprocessing.text.Tokenizer at 0x7bfa55895c00>,\n 'Exon': <keras.src.legacy.preprocessing.text.Tokenizer at 0x7bfa56905b40>,\n 'Feature': <keras.src.legacy.preprocessing.text.Tokenizer at 0x7bfa54247790>,\n 'Variant_Classification': <keras.src.legacy.preprocessing.text.Tokenizer at 0x7bfb740f34c0>}"},"metadata":{}}]},{"cell_type":"code","source":"import tensorflow as tf\nfrom tensorflow.keras.models import Model\nfrom tensorflow.keras.layers import Input, Embedding, Dense, Attention, Concatenate, GlobalAveragePooling1D, Lambda, Add\nimport numpy as np\n\nvocab_size_motif = len(tokenizers['Mutational_Motif'].word_index) + 1\nvocab_size_genomic_bin = len(tokenizers['Genomic_Bin'].word_index) + 1\nvocab_hugo = len(tokenizers['Hugo_Symbol'].word_index) + 1\n\nvocab_variant = len(tokenizers['Variant_Classification'].word_index) + 1\nvocab_feature = len(tokenizers['Feature'].word_index) + 1\nvocab_exon = len(tokenizers['Exon'].word_index) + 1\n\nembedding_dim = 512\nnum_classes = len(grouped_df['primary_site'].unique())\nmax_len = int(grouped_df['Mutational_Motif'].apply(len).quantile(0.75))\n\ndef create_model(embedding_dim, num_classes):\n    def positional_encoding(position, d_model):\n        angle_rads = np.arange(position)[:, np.newaxis] / np.power(10000, (2 * (np.arange(d_model)[np.newaxis, :] // 2)) / np.float32(d_model))\n        sines = np.sin(angle_rads[:, 0::2])\n        cosines = np.cos(angle_rads[:, 1::2])\n        pos_encoding = np.concatenate([sines, cosines], axis=-1)\n        pos_encoding = pos_encoding[np.newaxis, ...]\n        return tf.cast(pos_encoding, dtype=tf.float32)\n\n    # Wejścia\n    input_motif = Input(shape=(max_len,), name='Mutational_Motif')\n    input_genomic_bin = Input(shape=(max_len,), name='Genomic_Bin')\n    input_hugo = Input(shape=(max_len,), name='Hugo_Symbol')\n    # Wejścia 2\n    input_variant = Input(shape=(max_len,), name='Variant_Classification')\n    input_feature = Input(shape=(max_len,), name='Feature')\n    input_exon = Input(shape=(max_len,), name='Exon')\n    \n    \n    positional_encoding_layer = positional_encoding(max_len, embedding_dim)\n    \n    # main part\n    motif_embedding = Embedding(input_dim=vocab_size_motif, output_dim=embedding_dim)(input_motif)\n    genomic_bin_embedding = Embedding(input_dim=vocab_size_genomic_bin, output_dim=embedding_dim)(input_genomic_bin)\n    hugo_embedding = Embedding(input_dim=vocab_hugo, output_dim=embedding_dim)(input_hugo)\n    # second part\n    variant_embedding = Embedding(input_dim=vocab_variant, output_dim=embedding_dim)(input_variant)\n    feature_embedding = Embedding(input_dim=vocab_feature, output_dim=embedding_dim)(input_feature)\n    exon_embedding = Embedding(input_dim=vocab_exon, output_dim=embedding_dim)(input_exon)\n    \n    motif_embedding = Add()([Add()([motif_embedding, variant_embedding]), positional_encoding_layer])\n    genomic_bin_embedding = Add()([Add()([genomic_bin_embedding, feature_embedding]), positional_encoding_layer])\n    hugo_embedding = Add()([Add()([hugo_embedding, exon_embedding]), positional_encoding_layer])\n\n    \n    attention_layer = Attention()([motif_embedding, genomic_bin_embedding, hugo_embedding])\n\n    concat_layer = Concatenate()([motif_embedding, genomic_bin_embedding, hugo_embedding, attention_layer])\n\n    gap_layer = GlobalAveragePooling1D()(concat_layer)\n\n    output = Dense(num_classes, activation='softmax', name='project_short_name')(gap_layer)\n\n    model = Model(inputs=[input_motif, input_genomic_bin, input_hugo, input_variant, input_feature, input_exon], outputs=[output])\n\n    model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])\n\n    return model","metadata":{"execution":{"iopub.status.busy":"2024-08-28T14:53:16.349246Z","iopub.execute_input":"2024-08-28T14:53:16.350583Z","iopub.status.idle":"2024-08-28T14:53:16.410198Z","shell.execute_reply.started":"2024-08-28T14:53:16.350519Z","shell.execute_reply":"2024-08-28T14:53:16.408743Z"},"trusted":true},"execution_count":56,"outputs":[]},{"cell_type":"code","source":"num_classes = len(grouped_df['primary_site'].unique())","metadata":{"execution":{"iopub.status.busy":"2024-08-28T14:53:16.720975Z","iopub.execute_input":"2024-08-28T14:53:16.721634Z","iopub.status.idle":"2024-08-28T14:53:16.729097Z","shell.execute_reply.started":"2024-08-28T14:53:16.721573Z","shell.execute_reply":"2024-08-28T14:53:16.727401Z"},"trusted":true},"execution_count":57,"outputs":[]},{"cell_type":"code","source":"X_gene = X['Hugo_Symbol']\nX_genomic_bin = X['Genomic_Bin']\nX_motif = X['Mutational_Motif']\n\nX_exon = X['Exon']\nX_feature = X['Feature']\nX_variant = X['Variant_Classification']","metadata":{"execution":{"iopub.status.busy":"2024-08-28T14:53:17.059054Z","iopub.execute_input":"2024-08-28T14:53:17.059609Z","iopub.status.idle":"2024-08-28T14:53:17.068019Z","shell.execute_reply.started":"2024-08-28T14:53:17.059561Z","shell.execute_reply":"2024-08-28T14:53:17.066546Z"},"trusted":true},"execution_count":58,"outputs":[]},{"cell_type":"code","source":"import numpy as np\nimport tensorflow as tf\nfrom tensorflow.keras.models import Model\nfrom tensorflow.keras.layers import Input, Embedding, LSTM, Dense, Attention, Concatenate, Lambda, GlobalAveragePooling1D\nfrom sklearn.model_selection import KFold\nfrom sklearn.metrics import precision_score, recall_score\n\nkf = KFold(n_splits=5, shuffle=True, random_state=42)\nall_scores = []\nall_scores_top3 = []\nall_scores_top5 = []\nall_precisions = []\nall_recalls = []\n\n# Assuming y is already defined as:\ny = np.array(grouped_df['primary_site'].tolist())\n\nfor train_index, test_index in kf.split(X_motif):\n    X_train_gene, X_test_gene = X_gene[train_index], X_gene[test_index]\n    X_train_genomic_bin, X_test_genomic_bin = X_genomic_bin[train_index], X_genomic_bin[test_index]\n    X_train_motif, X_test_motif = X_motif[train_index], X_motif[test_index]\n\n    X_train_exon, X_test_exon = X_exon[train_index], X_exon[test_index]\n    X_train_feature, X_test_feature = X_feature[train_index], X_feature[test_index]\n    X_train_variant, X_test_variant = X_variant[train_index], X_variant[test_index]\n    \n    \n    y_train, y_test = y[train_index], y[test_index]\n\n    model = create_model(embedding_dim, num_classes)\n\n    model.fit([X_train_motif, X_train_genomic_bin, X_train_gene, X_train_variant, X_train_feature, X_train_exon], y_train, epochs=20, validation_split=0.1, batch_size=128, verbose=1)\n\n    scores = model.evaluate([X_test_motif, X_test_genomic_bin, X_test_gene, X_test_variant, X_test_feature, X_test_exon], y_test, verbose=0)\n    all_scores.append(scores[1])\n\n    predictions = model.predict([X_test_motif, X_test_genomic_bin, X_test_gene, X_test_variant, X_test_feature, X_test_exon])\n\n    y_pred_classes = np.argmax(predictions, axis=1)\n\n    # Calculate precision and recall\n    precision = precision_score(y_test, y_pred_classes, average='weighted')\n    recall = recall_score(y_test, y_pred_classes, average='weighted')\n\n    all_precisions.append(precision)\n    all_recalls.append(recall)\n\n    top_3 = np.argsort(predictions, axis=1)[:, -3:]\n    correct = np.sum([1 if y_test[i] in top_3[i] else 0 for i in range(len(y_test))])\n    top_3_accuracy = correct / len(y_test)\n\n    all_scores_top3.append(top_3_accuracy)\n\n    top_5 = np.argsort(predictions, axis=1)[:, -5:]\n    correct = np.sum([1 if y_test[i] in top_5[i] else 0 for i in range(len(y_test))])\n    top_5_accuracy = correct / len(y_test)\n\n    all_scores_top5.append(top_5_accuracy)\n\nprint(f'Srednia dokladnosc top1: {np.mean(all_scores)}')\nprint(f'Srednia dokladnosc top3: {np.mean(all_scores_top3)}')\nprint(f'Srednia dokladnosc top5: {np.mean(all_scores_top5)}')\nprint(f'Srednia precyzja: {np.mean(all_precisions)}')\nprint(f'Srednia recall: {np.mean(all_recalls)}')\n","metadata":{"execution":{"iopub.status.busy":"2024-08-28T14:53:17.315134Z","iopub.execute_input":"2024-08-28T14:53:17.316079Z"},"trusted":true},"execution_count":null,"outputs":[{"name":"stdout","text":"Epoch 1/20\n\u001b[1m95/95\u001b[0m \u001b[32m━━━━━━━━━━━━━━━━━━━━\u001b[0m\u001b[37m\u001b[0m \u001b[1m61s\u001b[0m 622ms/step - accuracy: 0.3281 - loss: 2.4783 - val_accuracy: 0.4892 - val_loss: 1.9385\nEpoch 2/20\n\u001b[1m30/95\u001b[0m \u001b[32m━━━━━━\u001b[0m\u001b[37m━━━━━━━━━━━━━━\u001b[0m \u001b[1m38s\u001b[0m 600ms/step - accuracy: 0.4879 - loss: 1.9026","output_type":"stream"}]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]}]}